validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,264 @@
+        let whoami = self.adapter.whoami();
+        let validator = self
+            .channel
+            .spec
+            .validators
+            .into_iter()
+            .find(|&v| v.id == whoami);
+        let auth_token = self.adapter.get_auth(validator.unwrap()).unwrap();
+
+        let url = format!(
+            "{}/events-aggregates?after={}",
+            self.validator_url,
+            after.timestamp()
+        );
+        let future = self
             .client
-            .get(format!("{}/last-approved", self.sentry_url))
+            .get(&url)
+            .header("authorization", auth_token.to_string())
             .send()
-            .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
+            .and_then(|mut res: Response| res.json::<EventAggregateResponse>());
 
         await!(future.compat())
     }
+}
+
+fn propagate_to(
+    auth_token: &str,
+    timeout: u32,
+    validator: &ValidatorDesc,
+    messages: &[String],
+) -> Result<(), reqwest::Error> {
+    // create client with timeout
+    let client = reqwest::Client::builder()
+        .timeout(Duration::from_secs(timeout.into()))
+        .build()?;
+    let url = validator.url.to_string();
+
+    let response: SuccessResponse = client
+        .post(&url)
+        .header(AUTHORIZATION, auth_token.to_string())
+        .json(messages)
+        .send()?
+        .json()?;
+
+    Ok(())
+}
+
+fn handle_http_error(e: reqwest::Error, url: &str) {
+    if e.is_http() {
+        match e.url() {
+            None => println!("No Url given"),
+            Some(url) => println!("Problem making request to: {}", url),
+        }
+    }
+    // Inspect the internal error and output it
+    if e.is_serialization() {
+        let serde_error = match e.get_ref() {
+            None => return,
+            Some(err) => err,
+        };
+        println!("problem parsing information {}", serde_error);
+    }
 
-    pub fn get_event_aggregates() {
+    if e.is_client_error() {
+        println!("erorr sending http request for validator {}", url)
+    }
 
+    if e.is_redirect() {
+        println!("server redirecting too many times or making loop");
     }
+}
+
+pub async fn all_channels(
+    sentry_url: &str,
+    adapter: impl Adapter + 'static,
+) -> Result<Vec<Channel>, ()> {
+    let validator = adapter.whoami();
+    let url = sentry_url.to_owned();
+    let first_page = await!(fetch_page(url.clone(), 0, validator.clone())).unwrap();
+    if first_page.total_pages < 2 {
+        Ok(first_page.channels)
+    } else {
+        let mut all: Vec<ChannelAllResponse> = await!(try_join_all(
+            (0..first_page.total_pages).map(|i| fetch_page(url.clone(), i, validator.clone()))
+        ))
+        .unwrap();
+        all.push(first_page);
+        let result_all: Vec<Channel> = all
+            .into_iter()
+            .flat_map(|ch| ch.channels.into_iter())
+            .collect();
+        Ok(result_all)
+    }
+}
+
+pub async fn fetch_page(
```

Comment:
We need to be careful what we expose. fetch_page is not part of the public API as it's only used in the `SentryApi`.
I saw quite some places like that so we need to review our API soon.

Reply:
perhaps that's a problem of the documentation? pages should be documented IMO

Reply:
It is meant as inner fn for `all_channels()` only, which should be public API itself(as it is) and I guess we can later even move it to the `SentryApi`.
But for `fetch_page` not sure if should be exposed if we need it only for `all_channels`?